src/zkbob/ZkBobCompoundingMixin.sol

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.15;
+
+import "@openzeppelin/contracts/interfaces/IERC4626.sol";
+import "./ZkBobPool.sol";
+
+/**
+ * @title ZkBobCompoundingMixin
+ */
+abstract contract ZkBobCompoundingMixin is ZkBobPool {
+    using SafeERC20 for IERC20;
+
+    uint256 public investedAssetsAmount;
+
+    struct YieldParams {
+        // ERC4626 vault address (or address(0) if not set)
+        address yield;
+        // expected amount of underlying tokens to be left at the pool after successful rebalance
+        uint96 buffer;
+        // operator address (or address(0) if permissionless)
+        address yieldOperator;
+        // slippage/rounding protection buffer, small part of accumulated interest that is non-claimable
+        uint96 dust;
+        // address to receive accumulated interest during the rebalance
+        address interestReceiver;
+        // maximum amount of underlying tokens that can be invested into vault
+        uint256 maxInvestedAmount;
+    }
+
+    YieldParams public yieldParams;
+
+    event UpdateYieldParams(YieldParams yieldParams);
+    event Claimed(address indexed yield, uint256 amount);
+    event Rebalance(address indexed yield, uint256 withdrawn, uint256 deposited);
+
+    // @inheritdoc ZkBobPool
+    function _withdrawToken(address _user, uint256 _tokenAmount) internal override {
+        uint256 underlyingBalance = IERC20(token).balanceOf(address(this));
+        if (underlyingBalance < _tokenAmount) {
+            (address yieldAddress, uint256 buffer) = (yieldParams.yield, yieldParams.buffer);
+            uint256 remainder = _tokenAmount - underlyingBalance;
+            uint256 investedAssets = investedAssetsAmount;
+            uint256 withdrawAmount = investedAssets > remainder + buffer ? remainder + buffer : investedAssets;
+            investedAssetsAmount = investedAssets - withdrawAmount;
+            IERC4626(yieldAddress).withdraw(withdrawAmount, address(this), address(this));
+            emit Rebalance(yieldAddress, withdrawAmount, 0);
+        }
+        IERC20(token).safeTransfer(_user, _tokenAmount);
+    }
+
+    /**
+     * @dev Updates yield parameters.
+     * Callable only by the contract owner / proxy admin.
+     * @param _yieldParams new yield parameters.
+     */
+    function updateYieldParams(YieldParams memory _yieldParams) external onlyOwner {
+        address yieldAddress = yieldParams.yield;
+        require(
+            _yieldParams.yield == yieldAddress || investedAssetsAmount == 0, "ZkBobCompounding: another yield is active"
+        );
+        require(
+            _yieldParams.yield == address(0) || _yieldParams.interestReceiver != address(0),
+            "ZkBobCompounding: zero interest receiver"
+        );
+
+        if (_yieldParams.yield != yieldAddress) {
+            if (_yieldParams.yield != address(0)) {
+                IERC20(token).approve(_yieldParams.yield, type(uint256).max);
```

Comment:
I think in order to properly limit any risk exposure at `maxInvestedAmount`, we need to get rid of an infinite approval. Let's make an exact approval only during an actual deposit as part of the rebalance.